### DISCUSSION TITLE
How do I use list.GetChangesBatchAsync() to Get ALL Changes for a Given Change Query?

### DISCUSSION THREAD

**Discussion by adamfisher** [pnp/pnpcore#1228] [Q&A] [answered]
I am trying to create an extension method that will allow me to get ALL of the changes that occurred for a given change query and I thought I would use `GetChangesBatchAsync()` to do it but I am confused how that is supposed to work to get everything? Won't subsequent calls need the change token? Just wondering how to build up the batch call to get the changes?

```csharp
public static async Task<IList<IChangeItem>> GetAllChangesAsync(this IList list, ChangeQueryOptions queryOptions)
{
	Batch batch = list.PnPContext.NewBatch()!;
	IEnumerableBatchResult<IChange>? changes = await list.GetChangesBatchAsync(batch, queryOptions);

	// Do something to get ALL changes...

	var results = await list.PnPContext.ExecuteAsync(batch);

	// How do I get all changes in one list to return?
}
}
```

**Comment by jansenbe** [comment]
@adamfisher : why do you need multiple calls to get all changes. If you use below you'll get all changes returned. Are you looking into paging the changes? The idea is you initially read the list information you need using regular APIs, then get the last change token (you can use the `LatestFirst` property of the `ChangeQueryOptions`) and store the change token. When you later on want to understand what happened you retrieve your stored change token and use that as a point in time to get the changes from.

```csharp
var changes = await list.GetChangesAsync(new ChangeQueryOptions(true, true));
```

**Reply by adamfisher** [reply]
I had over 1,000 changes I needed to process and while I could increase the [`FetchLimit`](https://pnp.github.io/pnpcore/api/PnP.Core.Model.SharePoint.ChangeQueryOptions.html#PnP_Core_Model_SharePoint_ChangeQueryOptions_FetchLimit) on the query options, I wasn't sure how I could set that. I just want to get all of the changes at once.

**Comment by adamfisher** [comment, marked as answer]
I ended up doing this since I couldn't figure out how to use the batch version:

```csharp
public static async Task<List<IChange>> GetAllChangesAsync(this IList list, ChangeQueryOptions queryOptions)
{
	IList<IChange> changes = await list.GetChangesAsync(queryOptions);
	List<IChange> results = new(changes);

	while (changes.Count > 0)
	{
		var lastChangeToken = changes.Last().ChangeToken;
		queryOptions.ChangeTokenStart = lastChangeToken;
		changes = await list.GetChangesAsync(queryOptions);
		results.AddRange(changes);
	}

	return results;
}
```

**Reply by jansenbe** [reply]
That's a good approach @adamfisher .
